sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs

```diff
@@ -0,0 +1,132 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class OperatorsShouldBeOverloadedConsistently : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4050";
+        private const string MessageFormat = "Provide an implementation for: {0}.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static class MethodName
+        {
+            public const string OperatorPlus = "operator+";
+            public const string OperatorMinus = "operator-";
+            public const string OperatorEquals = "operator==";
+            public const string OperatorNotEquals = "operator!=";
+
+            public const string ObjectEquals = "Object.Equals";
+            public const string ObjectGetHashCode = "Object.GetHashCode";
+        }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                if (classSymbol == null || classDeclaration.Identifier.IsMissing)
+                {
+                    return;
+                }
+
+                var classMethods = classSymbol
+                    .GetMembers()
+                    .OfType<IMethodSymbol>()
+                    .Where(m => !m.IsConstructor())
+                    .ToList();
+
+                var implementedMethods = new HashSet<string>();
+                implementedMethods.Add(GetNameOrDefault(MethodName.OperatorPlus, classMethods, KnownMethods.IsOperatorBinaryPlus));
```

Comment:
I would suggest something like calling a method returning `IEnumerable<string>` called `GetImplementedMethods(ISymbol classSymbol)` (or something like this) which will use the following pattern:
`if (methods.Any(KnownMethod.IsOperatorPlus)) return OperatorPlus;`

And in here you simply do `var implementedMethods = new HashSet<string>(GetImplementedMethods(classSymbol));`